pubsub/query/query.go

```diff
@@ -0,0 +1,202 @@
+	opGreater
+	opEqual
+	opContains
+)
+
+// Matches returns true if the query matches a given set of tags, false otherwise.
+//
+// For example, query "name=John" matches tags = {"name": "John"}. More
+// examples could be found in parser_test.go and query_test.go.
+func (q *Query) Matches(tags map[string]interface{}) bool {
+	if len(tags) == 0 {
+		return false
+	}
+
+	buffer, begin, end := q.parser.Buffer, 0, 0
+
+	var tag string
+	var op operator
+
+	for _, token := range q.parser.Tokens() {
+		switch token.pegRule {
+
+		case rulePegText:
+			begin, end = int(token.begin), int(token.end)
+		case ruletag:
+			tag = buffer[begin:end]
+		case rulele:
+			op = opLessEqual
+		case rulege:
+			op = opGreaterEqual
+		case rulel:
+			op = opLess
+		case ruleg:
+			op = opGreater
+		case ruleequal:
+			op = opEqual
+		case rulecontains:
+			op = opContains
+		case rulevalue:
+			if !match(tag, op, reflect.ValueOf(buffer[begin:end]), tags) {
+				return false
+			}
+		case rulenumber:
+			number := buffer[begin:end]
+			if strings.Contains(number, ".") {
+				value, err := strconv.ParseFloat(number, 32)
+				if err != nil {
+					panic(fmt.Sprintf("got %v. number format errors should be handled by the parser", err))
+				}
+				if !match(tag, op, reflect.ValueOf(value), tags) {
+					return false
+				}
+			} else {
+				value, err := strconv.Atoi(number)
+				if err != nil {
+					panic(fmt.Sprintf("got %v. number format errors should be handled by the parser", err))
+				}
+				if !match(tag, op, reflect.ValueOf(value), tags) {
+					return false
+				}
+			}
+		case ruletime:
+			value, err := time.Parse(time.RFC3339, buffer[begin:end])
+			if err != nil {
+				panic(fmt.Sprintf("got %v. time format errors should be handled by the parser", err))
+			}
+			if !match(tag, op, reflect.ValueOf(value), tags) {
+				return false
+			}
+		case ruledate:
+			value, err := time.Parse("2006-01-02", buffer[begin:end])
+			if err != nil {
+				panic(fmt.Sprintf("got %v. time format errors should be handled by the parser", err))
+			}
+			if !match(tag, op, reflect.ValueOf(value), tags) {
+				return false
+			}
+		}
+	}
+
+	return true
+}
+
+func match(tag string, op operator, operand reflect.Value, tags map[string]interface{}) bool {
+	value, ok := tags[tag]
+	if !ok { // no tag in tags matching a tag from query
+		return false
+	}
+	switch operand.Kind() {
+	case reflect.Struct: // time
+		v, ok := value.(time.Time)
+		if !ok {
+			return false
+		}
+		switch op {
+		case opLessEqual:
+			return v.Before(operand.Interface().(time.Time)) || v.Equal(operand.Interface().(time.Time))
+		case opGreaterEqual:
+			return v.Equal(operand.Interface().(time.Time)) || v.After(operand.Interface().(time.Time))
+		case opLess:
+			return v.Before(operand.Interface().(time.Time))
+		case opGreater:
+			return v.After(operand.Interface().(time.Time))
+		case opEqual:
+			return v.Equal(operand.Interface().(time.Time))
+		}
+	case reflect.Float64:
+		v, ok := value.(float64)
```

Comment:
I don't know how the parser works and all, but it would be good to have an end-to-end test to make sure matches don't fail on such an issue (maybe this is more a tendermint problem than this library)